bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -271,12 +261,17 @@ public void validate() {
        * {@link #startKey} and {@link #stopKey}.
        */
       @Override
-      public BoundedSource.BoundedReader<Result> createReader(PipelineOptions options)
+      public final BoundedSource.BoundedReader<Type> createReader(PipelineOptions options)
           throws IOException {
         Scan scan = new Scan(configuration.getScan());
         scan.setStartRow(startKey);
         scan.setStopRow(stopKey);
-        return new CloudBigtableIO.Source.Reader(this, configuration, scan);
+        return AbstractSource.this.createReader(scan);
+      }
+
+      @Override
+      public final Coder<Type> getDefaultOutputCoder() {
+        return AbstractSource.this.getDefaultOutputCoder();
```

Comment:
sign! this is the reason I don't like nested inner classes.
I totally lost track of what is what?
Why there are nested classes in AbstractSource? Why can't SourceWithKeys extends AbstractSource?